spec/ameba/rule/typing/method_parameter_type_restriction_spec.cr

```diff
@@ -0,0 +1,200 @@
+          expect_no_issues rule, <<-CRYSTAL
+            protected def hello(a)
+              "hello world"
+            end
+            CRYSTAL
+        end
+
+        it "fails if a public or private method doesn't have a return type restriction" do
+          expect_issue rule, <<-CRYSTAL
+            def hello(a)
+                    # ^ error: Method parameter should have a type restriction
+              "hello world"
+            end
+
+            private def hello(a)
+                            # ^ error: Method parameter should have a type restriction
+              "hello world"
+            end
+            CRYSTAL
+        end
+      end
+
+      context "#protected_methods" do
+        rule = MethodParameterTypeRestriction.new
+        rule.protected_methods = true
+
+        it "passes if a method has a parameter type restriction" do
+          expect_no_issues rule, <<-CRYSTAL
+            protected def hello(a : String) : String
+              "hello world" + a
+            end
+            CRYSTAL
+        end
+
+        it "passes if a private method param doesn't have a type restriction" do
+          expect_no_issues rule, <<-CRYSTAL
+            private def hello(a)
+              "hello world"
+            end
+            CRYSTAL
+        end
+
+        it "fails if a public or protected method doesn't have a parameter type restriction" do
+          expect_issue rule, <<-CRYSTAL
+            def hello(a)
+                    # ^ error: Method parameter should have a type restriction
+              "hello world"
+            end
+
+            protected def hello(a)
+                              # ^ error: Method parameter should have a type restriction
+              "hello world"
+            end
+            CRYSTAL
+        end
+      end
+
+      context "#default_value" do
+        it "fails if a method param with a default value doesn't have a type restriction" do
+          rule = MethodParameterTypeRestriction.new
+          rule.default_value = true
+
+          expect_issue rule, <<-CRYSTAL
+            def hello(a = "world")
+                    # ^ error: Method parameter should have a type restriction
+              "hello \#{a}"
+            end
+            CRYSTAL
+        end
+      end
+
+      context "#block_param" do
+        rule = MethodParameterTypeRestriction.new
+        rule.block_param = true
+
+        it "fails if a block param without a name doesn't have a type restriction" do
+          expect_issue rule, <<-CRYSTAL
+            def hello(&)
+                     # ^ error: Method parameter should have a type restriction
+              "hello"
+            end
+            CRYSTAL
+        end
+
+        it "fails if a block param with a name doesn't have a type restriction" do
+          expect_issue rule, <<-CRYSTAL
+            def hello(&a)
+                     # ^ error: Method parameter should have a type restriction
+              "hello, \#{a.call}"
```

Comment:
```suggestion
              "hello, #{a.call}"
```

Reply:
The backslash is needed as the interpolation should be happening in the parsed code, not the heredoc creating the code